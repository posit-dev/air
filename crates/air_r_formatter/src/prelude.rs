pub(crate) use crate::{AsFormat as _, FormatNodeRule, RFormatContext, RFormatter};
pub use biome_formatter::prelude::*;
pub use biome_rowan::{AstNode as _, AstNodeList as _, AstSeparatedList as _};
