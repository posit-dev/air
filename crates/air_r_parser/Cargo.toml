[package]
authors.workspace = true
categories.workspace = true
description = "Air R parser"
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
name = "air_r_parser"
publish = false
repository.workspace = true
rust-version.workspace = true
version = "0.0.0"

[dependencies]
air_r_factory = { workspace = true }
air_r_syntax = { workspace = true }
biome_diagnostics = { workspace = true }
biome_parser = { workspace = true }
biome_rowan = { workspace = true }
biome_unicode_table = { workspace = true }
serde = { workspace = true, features = ["derive"] }
tracing = { workspace = true }
tree-sitter = { workspace = true }
tree-sitter-r = { workspace = true }

[dev-dependencies]
biome_console = { workspace = true }
biome_diagnostics = { workspace = true }
insta = { workspace = true }
line_ending = { workspace = true }
tests_macros = { workspace = true }

# cargo-workspaces metadata
[package.metadata.workspaces]
independent = true

[lints]
workspace = true
