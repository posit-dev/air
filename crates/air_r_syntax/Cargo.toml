[package]
name = "air_r_syntax"
version = "0.0.0"
authors.workspace = true
edition.workspace = true
rust-version.workspace = true
description = "SyntaxKind and common rowan definitions for air_r_parser"
homepage.workspace = true
repository.workspace = true
license.workspace = true
keywords.workspace = true
categories.workspace = true
publish = false

[dependencies]
biome_rowan = { workspace = true, features = ["serde"] }
serde = { workspace = true, features = ["derive"] }

[lints]
workspace = true
