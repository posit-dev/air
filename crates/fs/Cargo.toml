
[package]
authors.workspace = true
categories.workspace = true
description = "Air filesystem utilities"
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
name = "fs"
publish = false
repository.workspace = true
rust-version.workspace = true
version = "0.0.0"

[dependencies]
path-absolutize = { workspace = true, features = [
    "once_cell_cache",
    "use_unix_paths_on_wasm",
] }

[lints]
workspace = true
