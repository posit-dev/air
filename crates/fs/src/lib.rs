use path_absolutize::Absolutize;
use std::ffi::OsStr;
use std::path::Path;
use std::path::PathBuf;

pub fn has_r_extension(path: &Path) -> bool {
    path.extension()
        .and_then(OsStr::to_str)
        .map_or(false, is_r_extension)
}

pub fn is_r_extension(extension: &str) -> bool {
    matches!(extension, "r" | "R")
}

/// Convert any path to an absolute path (based on the current working
/// directory).
pub fn normalize_path<P: AsRef<Path>>(path: P) -> PathBuf {
    let path = path.as_ref();
    if let Ok(path) = path.absolutize() {
        path.to_path_buf()
    } else {
        path.to_path_buf()
    }
}

/// Convert an absolute path to be relative to the current working directory.
pub fn relativize_path<P: AsRef<Path>>(path: P) -> String {
    let path = path.as_ref();

    #[cfg(target_arch = "wasm32")]
    let cwd = Path::new(".");
    #[cfg(not(target_arch = "wasm32"))]
    let cwd = path_absolutize::path_dedot::CWD.as_path();

    if let Ok(path) = path.strip_prefix(cwd) {
        return format!("{}", path.display());
    }
    format!("{}", path.display())
}
