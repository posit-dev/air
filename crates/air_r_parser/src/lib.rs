mod error;
mod options;
mod parse;

#[allow(unused)]
mod treesitter;

use air_r_factory::RSyntaxFactory;
pub use error::ParseError;
pub use options::RParserOptions;
pub use parse::parse;
pub use parse::parse_r_with_cache;
pub use parse::Parse;

use air_r_syntax::RLanguage;
use biome_parser::tree_sink::LosslessTreeSink;

pub(crate) type RLosslessTreeSink<'source> = LosslessTreeSink<'source, RLanguage, RSyntaxFactory>;
