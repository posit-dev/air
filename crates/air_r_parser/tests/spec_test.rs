use air_r_parser::ParseError;
use air_r_parser::{parse, RParserOptions};
use air_r_syntax::{RRoot, RSyntaxNode};
use biome_console::fmt::{Formatter, Termcolor};
use biome_console::markup;
use biome_diagnostics::display::PrintDiagnostic;
use biome_diagnostics::termcolor;
use biome_diagnostics::DiagnosticExt;
use biome_parser::prelude::ParseDiagnostic;
use biome_rowan::SyntaxNode;
use biome_rowan::SyntaxSlot;
use biome_rowan::{AstNode, SyntaxKind};
use std::fmt::Write;
use std::fs;
use std::path::Path;

#[derive(Copy, Clone)]
pub enum ExpectedOutcome {
    Pass,
    Fail,
    Undefined,
}

#[ignore]
#[test]
fn quick_test() {
    let code = "
    1$foo
    1$2";
    let options = RParserOptions::default();
    let parsed = parse(code, options);
    let root = RRoot::unwrap_cast(parsed.syntax());
    println!("{root:#?}");
}

pub fn run(test_case: &str, _snapshot_name: &str, test_directory: &str, outcome_str: &str) {
    let outcome = match outcome_str {
        "ok" => ExpectedOutcome::Pass,
        "error" => ExpectedOutcome::Fail,
        "undefined" => ExpectedOutcome::Undefined,
        _ => panic!("Invalid expected outcome {outcome_str}"),
    };

    let test_case_path = Path::new(test_case);

    let file_name = test_case_path
        .file_name()
        .expect("Expected test to have a file name")
        .to_str()
        .expect("File name to be valid UTF8");

    let content = fs::read_to_string(test_case_path)
        .expect("Expected test path to be a readable file in UTF8 encoding");

    // Normalize to Unix line endings
    let content = line_ending::normalize(content);

    let options = RParserOptions::default();
    let parsed = parse(&content, options);
    let root = RRoot::unwrap_cast(parsed.syntax());
    let formatted_ast = format!("{:#?}", root);

    let mut snapshot = String::new();
    writeln!(snapshot, "\n## Input\n\n```R\n{content}\n```\n\n").unwrap();

    writeln!(
        snapshot,
        r#"## AST

```
{formatted_ast}
```

## CST

```
{:#?}
```
"#,
        parsed.syntax()
    )
    .unwrap();

    if parsed.has_errors() {
        let diagnostics: Vec<ParseDiagnostic> = parsed
            .errors()
            .iter()
            .map(ParseError::diagnostic)
            .map(ParseDiagnostic::clone)
            .collect();

        let mut diagnostics_buffer = termcolor::Buffer::no_color();

        let termcolor = &mut Termcolor(&mut diagnostics_buffer);
        let mut formatter = Formatter::new(termcolor);

        for diagnostic in diagnostics {
            let error = diagnostic
                .clone()
                .with_file_path(file_name)
                .with_file_source_code(&content);

            formatter
                .write_markup(markup! {
                    {PrintDiagnostic::verbose(&error)}
                })
                .expect("failed to emit diagnostic");
        }

        let formatted_diagnostics =
            std::str::from_utf8(diagnostics_buffer.as_slice()).expect("non utf8 in error buffer");

        if matches!(outcome, ExpectedOutcome::Pass) {
            panic!("Expected no errors to be present in a test case that is expected to pass but the following diagnostics are present:\n{formatted_diagnostics}")
        }

        writeln!(snapshot, "## Diagnostics\n\n```").unwrap();
        snapshot.write_str(formatted_diagnostics).unwrap();

        writeln!(snapshot, "```\n").unwrap();
    }

    match outcome {
        ExpectedOutcome::Pass => {
            let missing_required = formatted_ast.contains("missing (required)");
            if missing_required
                || parsed
                    .syntax()
                    .descendants()
                    .any(|node: RSyntaxNode| node.kind().is_bogus())
            {
                panic!("Parsed tree of a 'OK' test case should not contain any missing required children or bogus nodes: \n {formatted_ast:#?} \n\n {formatted_ast}");
            }

            let syntax = parsed.syntax();
            if has_bogus_nodes_or_empty_slots(&syntax) {
                panic!("modified tree has bogus nodes or empty slots:\n{syntax:#?} \n\n {syntax}")
            }
        }
        ExpectedOutcome::Fail => {
            if !parsed.has_errors() {
                panic!("Failing test must have diagnostics");
            }
        }
        _ => {}
    }

    insta::with_settings!({
        prepend_module_to_snapshot => false,
        snapshot_path => &test_directory,
    }, {
        insta::assert_snapshot!(file_name, snapshot);
    });
}

/// This check is used in the parser test to ensure it doesn't emit
/// bogus nodes without diagnostics.
pub fn has_bogus_nodes_or_empty_slots<L: biome_rowan::Language>(node: &SyntaxNode<L>) -> bool {
    node.descendants().any(|descendant| {
        let kind = descendant.kind();
        if kind.is_bogus() {
            return true;
        }

        if kind.is_list() {
            return descendant
                .slots()
                .any(|slot| matches!(slot, SyntaxSlot::Empty { .. }));
        }

        false
    })
}
