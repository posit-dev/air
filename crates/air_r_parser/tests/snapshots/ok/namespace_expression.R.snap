---
source: crates/air_r_parser/tests/spec_test.rs
expression: snapshot
---
## Input

```R
a::b
a:::b

a::`b`
a:::`b`

a::"b"
a:::"b"

a::b()
a:::b()

a::b$c()
a:::b@c()

a::...
a:::...

...::a
...:::a

...::...
...:::...

a::..1
a:::..1

..1::a
..1:::a

..1::..1
..1:::..1

```


## AST

```
RRoot {
    bom_token: missing (optional),
    expressions: RExpressionList [
        RNamespaceExpression {
            left: RIdentifier {
                name_token: IDENT@0..1 "a" [] [],
            },
            operator: COLON2@1..3 "::" [] [],
            right: RIdentifier {
                name_token: IDENT@3..4 "b" [] [],
            },
        },
        RNamespaceExpression {
            left: RIdentifier {
                name_token: IDENT@4..6 "a" [Newline("\n")] [],
            },
            operator: COLON3@6..9 ":::" [] [],
            right: RIdentifier {
                name_token: IDENT@9..10 "b" [] [],
            },
        },
        RNamespaceExpression {
            left: RIdentifier {
                name_token: IDENT@10..13 "a" [Newline("\n"), Newline("\n")] [],
            },
            operator: COLON2@13..15 "::" [] [],
            right: RIdentifier {
                name_token: IDENT@15..18 "`b`" [] [],
            },
        },
        RNamespaceExpression {
            left: RIdentifier {
                name_token: IDENT@18..20 "a" [Newline("\n")] [],
            },
            operator: COLON3@20..23 ":::" [] [],
            right: RIdentifier {
                name_token: IDENT@23..26 "`b`" [] [],
            },
        },
        RNamespaceExpression {
            left: RIdentifier {
                name_token: IDENT@26..29 "a" [Newline("\n"), Newline("\n")] [],
            },
            operator: COLON2@29..31 "::" [] [],
            right: RStringValue {
                value_token: R_STRING_LITERAL@31..34 "\"b\"" [] [],
            },
        },
        RNamespaceExpression {
            left: RIdentifier {
                name_token: IDENT@34..36 "a" [Newline("\n")] [],
            },
            operator: COLON3@36..39 ":::" [] [],
            right: RStringValue {
                value_token: R_STRING_LITERAL@39..42 "\"b\"" [] [],
            },
        },
        RCall {
            function: RNamespaceExpression {
                left: RIdentifier {
                    name_token: IDENT@42..45 "a" [Newline("\n"), Newline("\n")] [],
                },
                operator: COLON2@45..47 "::" [] [],
                right: RIdentifier {
                    name_token: IDENT@47..48 "b" [] [],
                },
            },
            arguments: RCallArguments {
                l_paren_token: L_PAREN@48..49 "(" [] [],
                items: RArgumentList [],
                r_paren_token: R_PAREN@49..50 ")" [] [],
            },
        },
        RCall {
            function: RNamespaceExpression {
                left: RIdentifier {
                    name_token: IDENT@50..52 "a" [Newline("\n")] [],
                },
                operator: COLON3@52..55 ":::" [] [],
                right: RIdentifier {
                    name_token: IDENT@55..56 "b" [] [],
                },
            },
            arguments: RCallArguments {
                l_paren_token: L_PAREN@56..57 "(" [] [],
                items: RArgumentList [],
                r_paren_token: R_PAREN@57..58 ")" [] [],
            },
        },
        RCall {
            function: RExtractExpression {
                left: RNamespaceExpression {
                    left: RIdentifier {
                        name_token: IDENT@58..61 "a" [Newline("\n"), Newline("\n")] [],
                    },
                    operator: COLON2@61..63 "::" [] [],
                    right: RIdentifier {
                        name_token: IDENT@63..64 "b" [] [],
                    },
                },
                operator: DOLLAR@64..65 "$" [] [],
                right: RIdentifier {
                    name_token: IDENT@65..66 "c" [] [],
                },
            },
            arguments: RCallArguments {
                l_paren_token: L_PAREN@66..67 "(" [] [],
                items: RArgumentList [],
                r_paren_token: R_PAREN@67..68 ")" [] [],
            },
        },
        RCall {
            function: RExtractExpression {
                left: RNamespaceExpression {
                    left: RIdentifier {
                        name_token: IDENT@68..70 "a" [Newline("\n")] [],
                    },
                    operator: COLON3@70..73 ":::" [] [],
                    right: RIdentifier {
                        name_token: IDENT@73..74 "b" [] [],
                    },
                },
                operator: AT@74..75 "@" [] [],
                right: RIdentifier {
                    name_token: IDENT@75..76 "c" [] [],
                },
            },
            arguments: RCallArguments {
                l_paren_token: L_PAREN@76..77 "(" [] [],
                items: RArgumentList [],
                r_paren_token: R_PAREN@77..78 ")" [] [],
            },
        },
        RNamespaceExpression {
            left: RIdentifier {
                name_token: IDENT@78..81 "a" [Newline("\n"), Newline("\n")] [],
            },
            operator: COLON2@81..83 "::" [] [],
            right: RDots {
                value_token: DOTS@83..86 "..." [] [],
            },
        },
        RNamespaceExpression {
            left: RIdentifier {
                name_token: IDENT@86..88 "a" [Newline("\n")] [],
            },
            operator: COLON3@88..91 ":::" [] [],
            right: RDots {
                value_token: DOTS@91..94 "..." [] [],
            },
        },
        RNamespaceExpression {
            left: RDots {
                value_token: DOTS@94..99 "..." [Newline("\n"), Newline("\n")] [],
            },
            operator: COLON2@99..101 "::" [] [],
            right: RIdentifier {
                name_token: IDENT@101..102 "a" [] [],
            },
        },
        RNamespaceExpression {
            left: RDots {
                value_token: DOTS@102..106 "..." [Newline("\n")] [],
            },
            operator: COLON3@106..109 ":::" [] [],
            right: RIdentifier {
                name_token: IDENT@109..110 "a" [] [],
            },
        },
        RNamespaceExpression {
            left: RDots {
                value_token: DOTS@110..115 "..." [Newline("\n"), Newline("\n")] [],
            },
            operator: COLON2@115..117 "::" [] [],
            right: RDots {
                value_token: DOTS@117..120 "..." [] [],
            },
        },
        RNamespaceExpression {
            left: RDots {
                value_token: DOTS@120..124 "..." [Newline("\n")] [],
            },
            operator: COLON3@124..127 ":::" [] [],
            right: RDots {
                value_token: DOTS@127..130 "..." [] [],
            },
        },
        RNamespaceExpression {
            left: RIdentifier {
                name_token: IDENT@130..133 "a" [Newline("\n"), Newline("\n")] [],
            },
            operator: COLON2@133..135 "::" [] [],
            right: RDotDotI {
                value_token: DOTDOTI@135..138 "..1" [] [],
            },
        },
        RNamespaceExpression {
            left: RIdentifier {
                name_token: IDENT@138..140 "a" [Newline("\n")] [],
            },
            operator: COLON3@140..143 ":::" [] [],
            right: RDotDotI {
                value_token: DOTDOTI@143..146 "..1" [] [],
            },
        },
        RNamespaceExpression {
            left: RDotDotI {
                value_token: DOTDOTI@146..151 "..1" [Newline("\n"), Newline("\n")] [],
            },
            operator: COLON2@151..153 "::" [] [],
            right: RIdentifier {
                name_token: IDENT@153..154 "a" [] [],
            },
        },
        RNamespaceExpression {
            left: RDotDotI {
                value_token: DOTDOTI@154..158 "..1" [Newline("\n")] [],
            },
            operator: COLON3@158..161 ":::" [] [],
            right: RIdentifier {
                name_token: IDENT@161..162 "a" [] [],
            },
        },
        RNamespaceExpression {
            left: RDotDotI {
                value_token: DOTDOTI@162..167 "..1" [Newline("\n"), Newline("\n")] [],
            },
            operator: COLON2@167..169 "::" [] [],
            right: RDotDotI {
                value_token: DOTDOTI@169..172 "..1" [] [],
            },
        },
        RNamespaceExpression {
            left: RDotDotI {
                value_token: DOTDOTI@172..176 "..1" [Newline("\n")] [],
            },
            operator: COLON3@176..179 ":::" [] [],
            right: RDotDotI {
                value_token: DOTDOTI@179..182 "..1" [] [],
            },
        },
    ],
    eof_token: EOF@182..183 "" [Newline("\n")] [],
}
```

## CST

```
0: R_ROOT@0..183
  0: (empty)
  1: R_EXPRESSION_LIST@0..182
    0: R_NAMESPACE_EXPRESSION@0..4
      0: R_IDENTIFIER@0..1
        0: IDENT@0..1 "a" [] []
      1: COLON2@1..3 "::" [] []
      2: R_IDENTIFIER@3..4
        0: IDENT@3..4 "b" [] []
    1: R_NAMESPACE_EXPRESSION@4..10
      0: R_IDENTIFIER@4..6
        0: IDENT@4..6 "a" [Newline("\n")] []
      1: COLON3@6..9 ":::" [] []
      2: R_IDENTIFIER@9..10
        0: IDENT@9..10 "b" [] []
    2: R_NAMESPACE_EXPRESSION@10..18
      0: R_IDENTIFIER@10..13
        0: IDENT@10..13 "a" [Newline("\n"), Newline("\n")] []
      1: COLON2@13..15 "::" [] []
      2: R_IDENTIFIER@15..18
        0: IDENT@15..18 "`b`" [] []
    3: R_NAMESPACE_EXPRESSION@18..26
      0: R_IDENTIFIER@18..20
        0: IDENT@18..20 "a" [Newline("\n")] []
      1: COLON3@20..23 ":::" [] []
      2: R_IDENTIFIER@23..26
        0: IDENT@23..26 "`b`" [] []
    4: R_NAMESPACE_EXPRESSION@26..34
      0: R_IDENTIFIER@26..29
        0: IDENT@26..29 "a" [Newline("\n"), Newline("\n")] []
      1: COLON2@29..31 "::" [] []
      2: R_STRING_VALUE@31..34
        0: R_STRING_LITERAL@31..34 "\"b\"" [] []
    5: R_NAMESPACE_EXPRESSION@34..42
      0: R_IDENTIFIER@34..36
        0: IDENT@34..36 "a" [Newline("\n")] []
      1: COLON3@36..39 ":::" [] []
      2: R_STRING_VALUE@39..42
        0: R_STRING_LITERAL@39..42 "\"b\"" [] []
    6: R_CALL@42..50
      0: R_NAMESPACE_EXPRESSION@42..48
        0: R_IDENTIFIER@42..45
          0: IDENT@42..45 "a" [Newline("\n"), Newline("\n")] []
        1: COLON2@45..47 "::" [] []
        2: R_IDENTIFIER@47..48
          0: IDENT@47..48 "b" [] []
      1: R_CALL_ARGUMENTS@48..50
        0: L_PAREN@48..49 "(" [] []
        1: R_ARGUMENT_LIST@49..49
        2: R_PAREN@49..50 ")" [] []
    7: R_CALL@50..58
      0: R_NAMESPACE_EXPRESSION@50..56
        0: R_IDENTIFIER@50..52
          0: IDENT@50..52 "a" [Newline("\n")] []
        1: COLON3@52..55 ":::" [] []
        2: R_IDENTIFIER@55..56
          0: IDENT@55..56 "b" [] []
      1: R_CALL_ARGUMENTS@56..58
        0: L_PAREN@56..57 "(" [] []
        1: R_ARGUMENT_LIST@57..57
        2: R_PAREN@57..58 ")" [] []
    8: R_CALL@58..68
      0: R_EXTRACT_EXPRESSION@58..66
        0: R_NAMESPACE_EXPRESSION@58..64
          0: R_IDENTIFIER@58..61
            0: IDENT@58..61 "a" [Newline("\n"), Newline("\n")] []
          1: COLON2@61..63 "::" [] []
          2: R_IDENTIFIER@63..64
            0: IDENT@63..64 "b" [] []
        1: DOLLAR@64..65 "$" [] []
        2: R_IDENTIFIER@65..66
          0: IDENT@65..66 "c" [] []
      1: R_CALL_ARGUMENTS@66..68
        0: L_PAREN@66..67 "(" [] []
        1: R_ARGUMENT_LIST@67..67
        2: R_PAREN@67..68 ")" [] []
    9: R_CALL@68..78
      0: R_EXTRACT_EXPRESSION@68..76
        0: R_NAMESPACE_EXPRESSION@68..74
          0: R_IDENTIFIER@68..70
            0: IDENT@68..70 "a" [Newline("\n")] []
          1: COLON3@70..73 ":::" [] []
          2: R_IDENTIFIER@73..74
            0: IDENT@73..74 "b" [] []
        1: AT@74..75 "@" [] []
        2: R_IDENTIFIER@75..76
          0: IDENT@75..76 "c" [] []
      1: R_CALL_ARGUMENTS@76..78
        0: L_PAREN@76..77 "(" [] []
        1: R_ARGUMENT_LIST@77..77
        2: R_PAREN@77..78 ")" [] []
    10: R_NAMESPACE_EXPRESSION@78..86
      0: R_IDENTIFIER@78..81
        0: IDENT@78..81 "a" [Newline("\n"), Newline("\n")] []
      1: COLON2@81..83 "::" [] []
      2: R_DOTS@83..86
        0: DOTS@83..86 "..." [] []
    11: R_NAMESPACE_EXPRESSION@86..94
      0: R_IDENTIFIER@86..88
        0: IDENT@86..88 "a" [Newline("\n")] []
      1: COLON3@88..91 ":::" [] []
      2: R_DOTS@91..94
        0: DOTS@91..94 "..." [] []
    12: R_NAMESPACE_EXPRESSION@94..102
      0: R_DOTS@94..99
        0: DOTS@94..99 "..." [Newline("\n"), Newline("\n")] []
      1: COLON2@99..101 "::" [] []
      2: R_IDENTIFIER@101..102
        0: IDENT@101..102 "a" [] []
    13: R_NAMESPACE_EXPRESSION@102..110
      0: R_DOTS@102..106
        0: DOTS@102..106 "..." [Newline("\n")] []
      1: COLON3@106..109 ":::" [] []
      2: R_IDENTIFIER@109..110
        0: IDENT@109..110 "a" [] []
    14: R_NAMESPACE_EXPRESSION@110..120
      0: R_DOTS@110..115
        0: DOTS@110..115 "..." [Newline("\n"), Newline("\n")] []
      1: COLON2@115..117 "::" [] []
      2: R_DOTS@117..120
        0: DOTS@117..120 "..." [] []
    15: R_NAMESPACE_EXPRESSION@120..130
      0: R_DOTS@120..124
        0: DOTS@120..124 "..." [Newline("\n")] []
      1: COLON3@124..127 ":::" [] []
      2: R_DOTS@127..130
        0: DOTS@127..130 "..." [] []
    16: R_NAMESPACE_EXPRESSION@130..138
      0: R_IDENTIFIER@130..133
        0: IDENT@130..133 "a" [Newline("\n"), Newline("\n")] []
      1: COLON2@133..135 "::" [] []
      2: R_DOT_DOT_I@135..138
        0: DOTDOTI@135..138 "..1" [] []
    17: R_NAMESPACE_EXPRESSION@138..146
      0: R_IDENTIFIER@138..140
        0: IDENT@138..140 "a" [Newline("\n")] []
      1: COLON3@140..143 ":::" [] []
      2: R_DOT_DOT_I@143..146
        0: DOTDOTI@143..146 "..1" [] []
    18: R_NAMESPACE_EXPRESSION@146..154
      0: R_DOT_DOT_I@146..151
        0: DOTDOTI@146..151 "..1" [Newline("\n"), Newline("\n")] []
      1: COLON2@151..153 "::" [] []
      2: R_IDENTIFIER@153..154
        0: IDENT@153..154 "a" [] []
    19: R_NAMESPACE_EXPRESSION@154..162
      0: R_DOT_DOT_I@154..158
        0: DOTDOTI@154..158 "..1" [Newline("\n")] []
      1: COLON3@158..161 ":::" [] []
      2: R_IDENTIFIER@161..162
        0: IDENT@161..162 "a" [] []
    20: R_NAMESPACE_EXPRESSION@162..172
      0: R_DOT_DOT_I@162..167
        0: DOTDOTI@162..167 "..1" [Newline("\n"), Newline("\n")] []
      1: COLON2@167..169 "::" [] []
      2: R_DOT_DOT_I@169..172
        0: DOTDOTI@169..172 "..1" [] []
    21: R_NAMESPACE_EXPRESSION@172..182
      0: R_DOT_DOT_I@172..176
        0: DOTDOTI@172..176 "..1" [Newline("\n")] []
      1: COLON3@176..179 ":::" [] []
      2: R_DOT_DOT_I@179..182
        0: DOTDOTI@179..182 "..1" [] []
  2: EOF@182..183 "" [Newline("\n")] []

```
